### PR TITLE
Convert pusher databases to async/await.

### PR DESCRIPTION
**Pull request by clokep**
Converts a few databases to async/await:
* push rules
* push rule
* pusher

I stayed away from event push actions since I think other PRs are touching that at the moment.

### CODE REVIEW

File: synapse/rest/client/v1/push_rule.py [hunk 1]
```diff
@@ -163,7 +164,7 @@ def notify_user(self, user_id):
         stream_id, _ = self.store.get_push_rules_stream_token()
         self.notifier.on_new_event("push_rules_key", stream_id, users=[user_id])
 
-    def set_rule_attr(self, user_id, spec, val):
+    def set_rule_attr(self, user_id, spec, val) -> Awaitable:
```
Comment (erikjohnston) [inline synapse/rest/client/v1/push_rule.py:167]: Can we make this `async def` and await inside? I think that helps with stack traces

Comment (clokep) [reply synapse/rest/client/v1/push_rule.py:167]: Sure! We use this pattern quite a lot, unfortunately. Not sure if we will always want to change it or not. (It also makes the typing MUCH clearer IMO.)

Comment (erikjohnston) [reply synapse/rest/client/v1/push_rule.py:167]: Yeah, i think we almost always do. I think having `Awaitable` in the type signature is probably something we should view as code smell

Comment (clokep) [reply synapse/rest/client/v1/push_rule.py]: I think you just made a future PR of mine much larger. 😉

**Pull request merged**